src/bsoncxx/CMakeLists.txt

```
@@ -46,15 +47,23 @@ elseif(BSONCXX_POLY_OPTIONS_SET EQUAL 0)
         # If we are in C++17 mode, use the C++17 versions
         set(BSONCXX_POLY_USE_STD ON)
         message(STATUS "Auto-configuring bsoncxx to use C++17 std library polyfills since C++17 is active and user didn't specify otherwise")
+    elseif(ENABLE_BSONCXX_POLY_USE_IMPLS)
+        # If enabled, use bsoncxx implementations instead of external libraries.
+        set(BSONCXX_POLY_USE_IMPLS ON)
+        message(STATUS "Auto-configuring bsoncxx to use its own polyfill implementations since C++17 is inactive")
     elseif(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
         # Otherwise, since MSVC can't handle MNMLSTC, default to boost
         set(BSONCXX_POLY_USE_BOOST ON)
         message(STATUS "Auto-configuring bsoncxx to use boost std library polyfills since C++17 is inactive and compiler is MSVC")
+        message(DEPRECATION "Default selection of external polyfill libraries is deprecated and will be removed in an upcoming major release. Set ENABLE_BSONCXX_POLY_USE_IMPLS=ON to enable recommended behavior.")
     else()
         # Otherwise, we are on a platform that can handle MNMLSTC
         set(BSONCXX_POLY_USE_MNMLSTC ON)
         message(STATUS "Auto-configuring bsoncxx to use MNMLSTC for polyfills since C++17 is inactive")
+        message(DEPRECATION "Default selection of external polyfill libraries is deprecated and will be removed in an upcoming major release. Set ENABLE_BSONCXX_POLY_USE_IMPLS=ON to enable recommended behavior.")
     endif()
+elseif(BSONCXX_POLY_USE_MNMLSTC OR BSONCXX_POLY_USE_BOOST)
+    message(DEPRECATION "Support for external polyfill libraries is deprecated and will be removed in an upcoming major release. Set BSONCXX_POLY_USE_IMPLS=ON (pre-C++17) or set CMAKE_CXX_STANDARD=17 (or newer) to select recommended behavior.")
```

```suggestion
elseif(BSONCXX_POLY_USE_MNMLSTC OR BSONCXX_POLY_USE_BOOST OR BSONCXX_POLY_USE_STD_EXPERIMENTAL)
    message(DEPRECATION "Support for experimental and external polyfill libraries is deprecated and will be removed in an upcoming major release. Set BSONCXX_POLY_USE_IMPLS=ON (pre-C++17) or set CMAKE_CXX_STANDARD=17 (or newer) to select recommended behavior.")
endif()
```

Suggest also warning for the experimental polyfill.

---

The experimental std polyfill option currently has the following documentation:

> Note that this polyfill is _not_ recommended and is unsupported.

This line has been present [since April 2017](https://github.com/mongodb/mongo-cxx-driver/commit/612b31790c4d486a8124133355e0854c89ce8e48). I neglected to add a warning for it as I do not expect it is being used. I can still include it in the warnings if preferable.

On that note, do you think we could simply remove experimental std altogether, given it's been long-documented as unsupported?

---

> On that note, do you think we could simply remove experimental std altogether, given it's been long-documented as unsupported?

Yes. I expect `std::experimental` is OK to remove given it has been long-documented as unsupported. I also expect it is not being used. Though I do not think it is necessary in this PR. I agree with leaving out the warning.

---

Deferring removal to [CXX-2814](https://jira.mongodb.org/browse/CXX-2814).